### DISCUSSION TITLE
Getting an error on calling the Reload REST API

### DISCUSSION THREAD

**Discussion by rumen-vasilev** [ptarmiganlabs/butler#543] [Q&A] [answered]
Hello,

I am trying to reload an application via the REST API, but unfortunately I am getting an internal server error. Please take a look at the screenshot.

API Call via Windows Command Prompt:
![image](https://user-images.githubusercontent.com/83919259/183934391-13ab6946-6654-44d3-954b-65c84ccf491b.png)

Error log Butler:
![image](https://user-images.githubusercontent.com/83919259/183939454-e744b92d-48c9-40ed-92c9-32d87003db0b.png)

I am actually running the pre-build application (ver. 8.4.0) on a Windows 2019 machine. Does somebody have the same problem?

BR,
Rumen

**Comment by mountaindude** [comment]
Exactly how do you call Butler's API?
If you can share that it may be possible to test things here and possibly
reproduce the issue.

The starting-task-API is very flexible and powerful, but you need to pass
in parameters the correct way. The helper sub included in the repo and in
the demo app is prob easiest way to use that API from within a Sense script.

**Comment by mountaindude** [comment, marked as answer]
@rumen-vasilev 
I did a quick test with 8.4.2 (stand-alone butler.exe running on WinSrv 2016) here and for me the API works:

Using cURL(on macOS)  to call the API

```bash
➜  temp curl -X "PUT" "http://192.168.100.110:8081/v4/app/f42d6b16-8faf-45ca-a783-59f9da47db6d/reload" -H 'Content-Type: application/json; charset=utf-8' -d $'{}'

{"appId":"f42d6b16-8faf-45ca-a783-59f9da47db6d"}%
➜  temp
```


The result in the log is:

<img width="1212" alt="image" src="https://user-images.githubusercontent.com/1029262/184067303-906d87d6-8a6f-4a4b-9452-cd9facc41fd8.png">


The fact that it's Enigma that complains about "Access denied" makes me wonder if your Butler for some reason cannot connect to the Sense server. Or if there are permissions issues in the file system on the Windows server.


## Ideas for debugging the issue

### Ensure connectivity to Sense server

You can try the Butler Ping endpoint to rule out Butler-to-Sense connectivity issues:

Using cURL (on macOS again): 
`curl "http://192.168.100.110:8081/v4/butlerping"`

Looking good, which means Butler's communication with the Sense server is working. I.e. no certificate problems or similar.

<img width="482" alt="image" src="https://user-images.githubusercontent.com/1029262/184068004-6b8f2989-2110-47b7-9708-f9b79261a4c6.png">

The same can be done from PowerShell too. 
Note that you should do this from a computer where the QlikClient certificate is installed, for example one of the servers in the Sense cluster:

```powershell
Windows PowerShell
Copyright (C) 2016 Microsoft Corporation. All rights reserved.

PS C:\Users\goran> $url="http://192.168.100.110:8081/v4/butlerping"
PS C:\Users\goran> $cert = Get-ChildItem -Path "Cert:\CurrentUser\My" | Where-Object {$_.Subject -like '*QlikClient*'}
PS C:\Users\goran> Invoke-RestMethod -Uri $url -Method Get -Certificate $cert

response                  butlerVersion
--------                  -------------
Butler reporting for duty 8.4.2

PS C:\Users\goran>
```


### Test for file system permissions issues

Try running Butler as local admin. If things work then you probably have a file system permissions issue.

**Reply by rumen-vasilev** [reply]
Hello Göran,

curl "http://192.168.100.110:8081/v4/butlerping" works fine.
![image](https://user-images.githubusercontent.com/83919259/184150880-1ce7b5c3-0621-40df-ae22-5952c0029c9e.png)

I tried to call the api from the Butler demo app but also without success. I am getting the same error.
![image](https://user-images.githubusercontent.com/83919259/184157937-4bbb5f1d-4254-445c-b510-dfb795226ac2.png)

... and that is the debug log ...
![image](https://user-images.githubusercontent.com/83919259/184157237-66c67dcd-ee3e-4863-b091-2108b75bb8a5.png)

However I will talk with our administrator tomorrow. Maybe this is a firewall security issue at our site. I give you a feedback tomorrow. 

Thank you again for your support!

BR,
Rumen

**Comment by rumen-vasilev** [comment]
Hello Göran, I found the problem. It was because I tried to reload an application, which was published. That were the Sense security. :-)

**Comment by mountaindude** [comment]
Haha nice one!
Good that you found the cause of the issue. All good now thus?
